### ISSUE TITLE
The GET Catalog endpoint does not respect the "n" parameter

### ISSUE THREAD

**Issue by mssola** [distribution/distribution#1190]
This issue was reported in SUSE/Portus#548. I've investigated on this, and I don't know if at this point I'm just confused, or if there's something really wrong with this endpoint.

First of all, the `n` parameter. The documentation says that since getting the whole catalog can be expensive, this should be used to paginate the catalog. That's good, but a simple test tells me that this doesn't work. In the `registry/handlers/api_test.go` file I've added the following:

``` go
func TestCatalogAPIOVer100(t *testing.T) {
    env := newTestEnv(t, false)
    values := url.Values{}

    catalogURL, err := env.builder.BuildCatalogURL(values)
    if err != nil {
        t.Fatalf("unexpected error building catalog url: %v", err)
    }

    for i := 0; i < 110; i++ {
        name := fmt.Sprintf("image%d", i)
        createRepository(env, t, name, "sometag")
    }

    resp, err := http.Get(catalogURL)
    if err != nil {
        t.Fatalf("unexpected error issuing request: %v", err)
    }
    defer resp.Body.Close()

    checkResponse(t, "issuing catalog api check", resp, http.StatusOK)

    var ctlg struct {
        Repositories []string `json:"repositories"`
    }

    dec := json.NewDecoder(resp.Body)
    if err = dec.Decode(&ctlg); err != nil {
        t.Fatalf("error decoding fetched manifest: %v", err)
    }

    fmt.Printf("%#v\n", ctlg.Repositories)
}
```

As you can see, I just upload 110 repositories and then ask for the catalog _without_ specifying the `n` parameter. The result is that I get exactly 100 repositories instead of the expected 110. After investigating, I saw this in the `GetCatalog` function inside of `registry/handlers/catalog.go` file:

``` go
    maxEntries, err := strconv.Atoi(q.Get("n"))
    if err != nil || maxEntries < 0 {
        maxEntries = maximumReturnedEntries
    }
```

This is wrong, because if you don't specify the `n` parameter (as in "give me all you've got"), then `q.Get("n")` will return an empty string, and therefore `strconv.Atoi` will fail. Because of this, you always end up getting a fixed amount of 100 repositories, even if you wanted all the catalog, regardless of performance issues.

This is a bug, but there's something a bit more subtle but fundamentally flawed. The `GetCatalog` function ends up calling:

``` go
filled, err := ch.App.registry.Repositories(ch.Context, repos, lastEntry)
```

This, in turn, goes to the `Repositories` function that you can find in the `registry/storage/catalog.go` file. After looking at the code I feared that there was something wrong with it, so to confirm it I added a print statement to get the contents of `foundRepos` after the Walk call. The result: `foundRepos` contains _all_ the repositories. Thus, I don't know why do we even care about the `n` parameter if this function already walks through all the stored images anyways and then crops it to the demanded number of images :/

So we have two ways to fix this:

1 Remove the `n` parameter and pagination for this endpoint (something that admittedly we don't want because it breaks backwards compatibility, but it would be the easiest solution).
2 Improve the `Repositories` function so it only just walks through the stored images until `n` is met. This would be difficult to tackle, because the `last` parameter relies on the list of stored images to be sorted if I read it correctly (thus the `sort` after the `Walk` call).

What do you think ?

**Comment by stevvooe** [comment]
@mssola Sorry for not getting back to you earlier. I spoke with @flavio at DockerCon who had mentioned this. There is a lot going on here, so please bear with me.

The first thing to point out it that this endpoint is only intended for administrative functionality, such as search index syncing, so calling it as part of a command line, without caching results, is way outside of the intended or recommended use case. Fortunately, the API is designed to scale (and it will), while just the implementation is non-optimal. Most of this can be addressed with the right fixes.

`n` should limit the size of the response, without assigning rank to a dataset of unknown size only if present. API compliant clients should actually handle this by following the `Link` header, provided in the response. I took a peak at the code in the portus project and it doesn't seem to be doing this properly.

@pdevine Please investigate. I'm a little uncertain what is used in DTR. I'm guessing we properly follow the link header.

 [`Namespace.Repositories`](https://godoc.org/github.com/docker/distribution#Namespace) doesn't need another argument. `n` should set the size of the slice passed into this function. This happens [here](https://github.com/docker/distribution/blob/master/registry/handlers/catalog.go#L45).

> This, in turn, goes to the Repositories function that you can find in the registry/storage/catalog.go file. After looking at the code I feared that there was something wrong with it, so to confirm it I added a print statement to get the contents of foundRepos after the Walk call. The result: foundRepos contains all the repositories. Thus, I don't know why do we even care about the n parameter if this function already walks through all the stored images anyways and then crops it to the demanded number of images :/

This is clearly stated in #947 and several other issues. With proper performance improvements, this is just an implementation detail.

> 2 Improve the Repositories function so it only just walks through the stored images until n is met. This would be difficult to tackle, because the last parameter relies on the list of stored images to be sorted if I read it correctly (thus the sort after the Walk call).

I'd not recommend looking at both sides of that interface when doing this analysis. The abstract representation of repositories is a sorted set of names. One can very efficiently partition this into consistent, fixed-sized blocks (pages) of responses when splitting by key. Effectively, you always get `n` results without repeats from page to page, without causing shifts on insertion.

Let's optimize before considering API changes. Because we have an inefficient backend implementation of sorted set should not inform the change of this interface or the API. We simply need to make the backend faster. This could be as easy as syncing the repository data set into a redis sorted set and issuing catalog queries from there. Faster walk implementation, by shuttling past unneeded entries or with a find command on storage driver, would also address this.

Part of the problem is this [statement](https://github.com/docker/distribution/blob/master/registry/storage/catalog.go#L30):

```
// Walk each of the directories in our storage.  Unfortunately since there's no
// guarantee that storage will return files in lexigraphical order, we have
// to store everything another slice, sort it and then copy it back to our
// passed in slice.
```

This is just plain wrong. A simple sort call [here](https://github.com/docker/distribution/blob/master/registry/storage/walk.go#L29) would fix that.

In summary:

1. When `n` is not present, entire dataset should be returned. Specification leaves this up to registry, so clients should really handle link header, as stated in the api specification.
2. `Walk` should guarantee lexicographical order. That is literally the definition of walk (see [`filepath.Walk`](https://golang.org/pkg/path/filepath/#Walk). Many optimizations are not possible if walk isn't ordered. Ideally, the backend should actually guarantee sorted listing, but we can sort at each walk level.
3. `Repositories` in catalog.go should exit when the response slice is full, leveraging lexicographic ordering provided by correctly implement walk function.
4.  `Repositories` in catalog.go should use `ErrSkip` to control shuttling during walk. For example, if last="b/c", `Walk` should be instructed not to descend _`k`_ < "b", since we know that they won't be included in the result set. More generally, if _`k`_ < `last`, `ErrSkip` should be returned. If _`k`_ has children, partition children by `last` and only descend those after `last`. (It may already achieve this).
5. Catalog should synced and cached from backend.

If the above is covered, even just 1-4, we can have a performant, efficient, API compliant catalog traversal with no actual changes to the API.

**Comment by mssola** [comment]
> The first thing to point out it that this endpoint is only intended for administrative functionality, such as search index syncing, so calling it as part of a command line, without caching results, is way outside of the intended or recommended use case. Fortunately, the API is designed to scale (and it will), while just the implementation is non-optimal. Most of this can be addressed with the right fixes.

Yes, I know. In Portus we use it in a job used for syncing. There's also a command line tool, but that's only used for testing purposes ;)

> n should limit the size of the response, without assigning rank to a dataset of unknown size only if present. API compliant clients should actually handle this by following the Link header, provided in the response. I took a peak at the code in the portus project and it doesn't seem to be doing this properly.

Fixed with SUSE/Portus#602. Thanks for pointing this out :)

> This is clearly stated in #947 and several other issues. With proper performance improvements, this is just an implementation detail.

I see. Thanks for letting me know.

> Let's optimize before considering API changes. Because we have an inefficient backend implementation of sorted set should not inform the change of this interface or the API. We simply need to make the backend faster. This could be as easy as syncing the repository data set into a redis sorted set and issuing catalog queries from there. Faster walk implementation, by shuttling past unneeded entries or with a find command on storage driver, would also address this.

I totally agree.

> If the above is covered, even just 1-4, we can have a performant, efficient, API compliant catalog traversal with no actual changes to the API.

Good summary. This should be enough. Thanks for taking the time :)

**Comment by stevvooe** [comment]
@mssola No problem. I'm mostly interested in breaking this down so community members who are interested in making this endpoint more performant have the right direction.

If you or one of your team members could take it on, it would be great help. If not, I'll try to get this addressed internally, but I am not sure it is affecting us.

**Comment by pdevine** [comment]
I've been working on a fix for #1 and just need to do some testing before I send out a PR.  If #2 is fixed, I can look at a refactor for #3 and #4, although I'm slightly time constrained right now.

**Comment by stevvooe** [comment]
@pdevine Thanks. I'll get PR for item 2.

Although, I have found yet another performance issue in the implementation of `storage.Walk`. For every single entry for a given directory, [`StorageDriver.Stat` is called](https://github.com/docker/distribution/blob/master/registry/storage/walk.go#L30). In turn, for the s3 driver, list is called in `StorageDriver.Stat` to see if an entry is a directory. Basically, for drivers with an expensive stat call, this is very slow.

This may take some reworking of the storage driver API to work out correctly.

**Comment by stevvooe** [comment]
@pdevine Submitted #1227. Should open up item 3 and 4.

I'm not sure if any of this matters in the face of the inner loop stat call...

**Comment by pdevine** [comment]
I've just submitted PR #1241 which should fix issue 3.  The fourth issue was actually done as part of the original catalog implementation, but I've also added a way of short circuiting the entire walk when the return array is full and can't add any more repositories.  Once the PR is merged, I think we can safely close this issue.

**Comment by pdevine** [comment]
@stevvooe @RichardScothern I think we can safely close this issue now since items 1 through 4 are done.  Item 5 seems a little out of scope.

**Comment by stevvooe** [comment]
> Although, I have found yet another performance issue in the implementation of storage.Walk. For every single entry for a given directory, StorageDriver.Stat is called. In turn, for the s3 driver, list is called in StorageDriver.Stat to see if an entry is a directory. Basically, for drivers with an expensive stat call, this is very slow.

@pdevine I think item 6, unspecified in that list, is probably the largest culprit.

@mssola Could you give the current master a try and see if it helps with performance woes?

**Comment by dmp42** [comment]
@mssola any update on this?

**Comment by milosgajdos** [comment]
@Jamstah IIRC this is now fixed, right?

**Comment by Jamstah** [comment]
Yep
